README.md
<div id="top"></div>

<!-- PROJECT LOGO -->
<br>
<div align="center">
  <a href="https://bitbucket.org/robertofierimonte/avocados-technical-task/">
    <img src="docs/images/logo.png" alt="Logo" height="350">
  </a>

<h2 align="center">Avocados Technical Task</h2>
  <p>Technical task about building a simple ML model in Numpy and a Flask Rest API for CRUD operations.<br>
  Author: <a href="mailto:roberto.fierimonte@gmail.com"><b>Roberto Fierimonte</b></a></p>
</div>

<!-- TABLE OF CONTENTS -->
 <h3 align="center">Table of Contents</h3>
 <ol>
 <li>
  <a href="#about-the-project">About the Project</a>
 </li>
 <li>
  <a href="#getting-started">Getting Started</a>
  <ul>
   <li><a href="#prerequisites">Prerequisites</a></li>
   <li><a href="#repo-structure">Repo structure</a></li>
   <li><a href="#local-setup">Local setup</a></li>
   <li><a href="#container-orchestration">Container orchestration</a></li>
  </ul>
 </li>
 <li><a href="#using-the-api">Using the API</a></li>
 <li><a href="#running-the-ml-model">Running the ML model</a></li>
</ol>

<!-- ABOUT THE PROJECT -->
## About the Project

Generic takehome task for a Data Science / Machine Learning Engineer position.

<u>Requirements</u>:

1. Create a linear regression model estimating the function `y=f(x)` that represents the price of Avocados based on historical data. Furthermore, information related to the features that most drive this change in price should be pointed out.

    This should be done using only numpy and basic python - i.e not using higher-level packages. Basic machine learning consideration when preprocessing and handling data need to be taken into consideration. Lines of code should be commented thoroughly to show understanding.

    [Link to dataset](​https://drive.google.com/file/d/1rhRzA2s44I8ASm_bMHnCpmAz_mNJQ7M3/view?usp=sharing)

2. Using python and Flask construct a REST API about recipes.

    a. Create a few tables in a mysql database that has recipe information and recipe ingredients.

    b. Create an API to:

      * Get, edit, delete and create recipes.
      * Get recipes by certain `ingredient_name`

    <br/> Good python techniques should be used, especially regarding the design of the API. Furthermore, some error handling is also expected.

Please note that the two tasks are independent and can be tested and executed independently.

<p align="right">(<a href="#top">back to top</a>)</p>

## Getting Started

### Prerequisites
- Python 3.9 (we are using version `3.9.7`)
- [poetry](https://python-poetry.org/) (we are using version `1.2.2`)
- [pre-commit](https://pre-commit.com/)
- [Docker](https://www.docker.com/) (must have suppport for Docker Compose `v3` - we are using version `20.10.21`)
- [MySQL](https://www.mysql.com/) (optional for local testing - we are using version `8.0.31`)

### Repo structure

```
avocados-technical-task
├── api_examples
├── containers
├── data
├── docs
├── model
├── notebooks
├── scripts
└── src
     ├── api
     └── base
```

- The `api_examples` folder contains some `.json` files that can be used to test the API
- The `containers` folder contains the Dockerfile used to build the image for the Flask App
- The `data` folder is empty and it will be populated with the dataset after completing the local setup
- The `docs` folder contains the documentation about the project
- The `model` folder is empty and it will be populated with the serialised ML model files after running the training notebook
- The `notebooks` folder contains the notebooks for exploratory analysis over the data, as well as the training and testing of the model
- The `scripts` subfolder contains the script to initialise the MySQL database, as well as a function to clean the recipe name before feeding it to the API
- The `src` folder contains the source code for the project:
   - The `api` subfolder contains the code to run the Flask App
   - The `base` subfolder contains the code used for the exploratory analysis, data preprocessing, and ML model

### Local setup

In the repository, execute:

1. Install poetry if not available: `pip install poetry==1.2.2`
2. Install the required dependencies: `poetry install`
3. Install pre-commit hooks: `poetry run pre-commit install`
4. Download the dataset by running `make download-data`
5. Follow the instructions in the `.env.example` to set up the necessary environment variables
6. Run `make help` to see all the options provided in the Makefile

To run the code locally (on your machine):

7. Run `make mysql-local-setup` to initialise the MySQL database
8. Run `make run-server-local` to run the Flask App

Reference to section <a href="#using-the-api">Using the API</a> for more info on how to use the API.

<p align="right">(<a href="#top">back to top</a>)</p>

### Container Orchestration

Alongside running the API locally we provide the option to use Docker compose to host both the API and the MySQL database on containers. We do this through [Docker Compose](https://docs.docker.com/compose/). You can find the configuration for the systems in the [docker-compose.yml](./docker-compose.yml) file.

We defined two containers to be used in this application - the first container hosts the MySQL database and the second container hosts the Flask App. The App container is dependent on the MySQL container to be healthy before being spun up. Also the two containers will share a network. The Docker application can be built by running `make build-compose` (before starting it up for the first time or when changes are made to it), started by running `make run-compose`, and stopped by running `docker compose down`. After starting the application for the first time (or if you have deleted the database volume - see below) run `make mysql-setup-docker` to instantiate the database, tables, and triggers. This steps solves the issue of the syntax for MySQL triggers causing an error when trying to initiate the database using the native docker entrypoint.

The MySQL container uses a volume to persist the database, so that when the containers are shut down the database is restore and available for future sessions. To delete the volume at the same time of stopping the application, run `docker compose down -v`. **N.B.: this will remove all the tables** and you'll have to re-create them using the `make mysql-setup-docker` command before being able to use the API again.

<p align="right">(<a href="#top">back to top</a>)</p>

### Using the API

The API provides the following entry-points:

| HTTP Method | Entrypoint                               | Make Command                                                  | Notes                                            | Response codes             |
|-------------|------------------------------------------|---------------------------------------------------------------|--------------------------------------------------|----------------------------|
| GET         | `/recipes`                               | `make get-all-recipes`                                        |                                                  | 200 Success                |
| GET         | `/recipes/<recipe-name>`                 | `make get-recipe-by-name recipe=<recipe-name>`                |                                                  | 200 Success, 404 Not Found |
| POST        | `/recipes/<recipe-name>`                 | `make post-recipe-by-name recipe=<recipe-name>`               | Need `api_examples/post_<recipe-name>.json` file | 201 Created, 400 Bad Request, 409 Conflict |
| PUT         | `/recipes/<recipe-name>`                 | `make put-recipe=by-name recipe=<recipe-name>`                | Need `api_examples/put_<recipe-name>.json` file  | 200 Success, 400 Bad Request, 404 Not Found |
| DELETE      | `/recipes/<recipe-name>`                 | `make delete-recipe-by-name recipe=<recipe=name>`             |                                                  | 200 Success                |
| GET         | `/ingredients`                           | `make get-all-ingredients`                                    |                                                  | 200 Success                |
| GET         | `/ingredients/<ingredient-name>/recipes` | `make get-recipes-by-ingredient ingredient=<ingredient-name>` |                                                  | 200 Success, 404 Not Found |

A description for all the previous commands is also available by running `make help`.

When running the commands from the Makefile, first the `<recipe-name>` and the `<ingredient-name>` will be cleaned by making them lowercase, removing all alphabetical characters, and replacing whitespaces with underscores. This is done by running the [clean_string.py](./scripts/clean_string.py) script on the string. Afterwards, in case of POST or PUT requests related to a recipe, a `.json` file will be looked in the `api_examples` folder corresponding to the cleaned `<recipe-name>` and that will be passed as a payload to the API call.

For example, when running the command `make post-recipe-by-name recipe="Risotto gorgonzola, pears, and walnuts"`, the `<recipe-name>` will be cleaned to become `risotto_gorgonzola_pears_walnuts`, a file `post_risotto_gorgonzola_pears_walnuts.json` will be looked up inside [the `api_examples` folder](./api_examples/) and it will be used as the payload for the request to create the new recipe.

Of course you can also make a request pointing to a different file or alternatively using a `json` payload directly from Postman.

<p align="right">(<a href="#top">back to top</a>)</p>

### Running the ML Model

A trained model is not provided with this repository. To train the model please the [training notebook](notebooks/1-eda-model.ipynb) end to end. That will save the model in the [model](model/) folder for it to be used to predict on new data.

<p align="right">(<a href="#top">back to top</a>)</p>


CHANGELOG.md
## v1.0.0 (2022-12-12)

### Feat

- **src/api/**: Implemented CRUD Rest API for recipes
- **src/api/**: Fixed POST method for creating a request, as well as GET methods for all recipes and all ingredients
- First working version of POST API for recipes
- Improvements to the Recipes API
- **src/base/model.py**: Added methods to save and load model
- **src/api/**: Added initial version of Flask API
- Added MySQL queries to create new user and set up `recipes` schema
- **notebooks/1-eda-model.ipynb**: Added notebook for EDA and model training and testing
- **src/base/**: Added source code for Linear Regression model and data preprocessing

### Fix

- **notebooks/1-eda-model.ipynb**: Fixed a bug in model training that was using the target variable as input


src/api/app.py
import os

from loguru import logger
from flask import Flask, render_template, request, jsonify

from src.base.utils import hash_string, clean_string
from src.api.database import db, ma
from src.api.models import (
    Recipe,
    RecipeSchema,
    Ingredient,
    IngredientSchema,
    RecipeIngredients,
)


# Read the environment variables
mysql_user = os.environ.get("MYSQL_USER")
mysql_pwd = os.environ.get("MYSQL_PASSWORD")
mysql_host = os.environ.get("MYSQL_HOST", "localhost")
mysql_port = os.environ.get("MYSQL_PORT", 3306)
mysql_db = os.environ.get("MYSQL_DATABASE", "recipes")

# Create the app
app = Flask(__name__)
app.config[
    "SQLALCHEMY_DATABASE_URI"
] = f"mysql+pymysql://{mysql_user}:{mysql_pwd}@{mysql_host}:{mysql_port}/{mysql_db}"
app.config["SQLALCHEMY_ECHO"] = True

# Create the extension and initialise the app with the extension
db.init_app(app)
ma.init_app(app)

# Initialise the schema objects
recipe_schema = RecipeSchema()
recipes_schema = RecipeSchema(many=True)
ingredient_schema = IngredientSchema()
ingredients_schema = IngredientSchema(many=True)


@app.route("/")
def hello():
    """Root page method."""
    return render_template("index.html")


@app.route("/recipes", methods=["GET"])
def get_recipes():
    """Get all the recipes."""
    all_recipes = Recipe.query.all()
    result = recipes_schema.dump(all_recipes)
    return jsonify(result), 200


@app.route("/recipes/<recipe_name>", methods=["GET", "PUT", "POST", "DELETE"])
def recipe(recipe_name: str):
    """Create, get, update, or delete a recipe by its name.

    Args:
        recipe_name (str): Name of the recipe
    """
    # Get the name of the recipe and the corresponding ID
    logger.debug(f"Recipe Name: {recipe_name} .")
    recipe_id = hash_string(clean_string(recipe_name), 18)
    logger.debug(f"Recipe ID: {recipe_id} .")

    # If POST, create a new recipe
    if request.method == "POST":
        # If the recipe does not already exist, create the recipe, then check
        # all the ingredients.
        if Recipe.query.get(recipe_id) is None:
            try:
                ingredients = request.json.pop("ingredients")
                recipe = Recipe(id=recipe_id, **request.json)
                db.session.add(recipe)
                # If any of the ingredients does not exist, create it and set
                # the ref_unit_of_measure to the UOM specified in the recipe.
                # Finally, add the recipe and the ingredients to the mapping
                # table
                for ing in ingredients:
                    ing_name = ing["name"]
                    ing_id = hash_string(clean_string(ing_name), 18)
                    logger.debug(
                        f"Ingredient Name: {ing_name}, "
                        f"Cleaned Ingredient Name: {clean_string(ing_name)}, "
                        f"Ingredient ID: {ing_id} ."
                    )
                    ing_uom = ing["unit_of_measure"]
                    ing_qty = ing["quantity"]
                    ingredient = Ingredient(
                        id=ing_id, name=ing_name, ref_unit_of_measure=ing_uom
                    )
                    if Ingredient.query.get(ing_id) is None:
                        db.session.add(ingredient)
                    # Add the recipe and the ingredient to the mapping table
                    recipe_ingredient = RecipeIngredients(
                        recipe_id=recipe_id,
                        ingredient_id=ing_id,
                        unit_of_measure=ing_uom,
                        quantity=ing_qty,
                    )
                    db.session.add(recipe_ingredient)
                # Commit all the changes to the database
                db.session.commit()
                return recipe_schema.jsonify(recipe), 201
            # If the ingredients are not in the request, return an error (400)
            except KeyError as e:
                jsonify(
                    {"error:": f"Attribute {e} does not exist in the payload."}
                ), 400
        # If the recipe exists already, return an error (409)
        else:
            return jsonify({"error": "This recipe already exists."}), 409

    # If GET, get the recipe or return 404 Not Found
    elif request.method == "GET":
        recipe = db.get_or_404(Recipe, recipe_id)
        return recipe_schema.jsonify(recipe), 200

    # If PUT, update the recipe
    elif request.method == "PUT":
        # Get the recipe or return 404 Not Found
        recipe = db.get_or_404(Recipe, recipe_id)
        # If the list of ingredients has changed, update the ingredients. The
        # ingredients can be added, modified, or removed (using a special
        # keyword)
        ingredients = request.json.pop("ingredients", None)
        if ingredients is not None:
            try:
                for ing in ingredients:
                    ing_name = ing.pop("name")
                    ing_id = hash_string(clean_string(ing_name), 18)
                    logger.debug(
                        f"Ingredient Name: {ing_name}, "
                        f"Cleaned Ingredient Name: {clean_string(ing_name)}, "
                        f"Ingredient ID: {ing_id} ."
                    )
                    # If the ingredient is not part of the recipe anymore,
                    # delete it from the mapping table
                    if "quantity" in ing and ing["quantity"] == 0:
                        recipe_ingredient = RecipeIngredients.query.get_or_404(
                            (recipe_id, ing_id)
                        )
                        db.session.delete(recipe_ingredient)
                    # Otherwise...
                    else:
                        # If the ingredient is completely new, add it to the list
                        # of ingredients
                        if Ingredient.query.get(ing_id) is None:
                            ingredient = Ingredient(
                                id=ing_id,
                                name=ing_name,
                                ref_unit_of_measure=ing["unit_of_measure"],
                            )
                            db.session.add(ingredient)
                        # If the ingredient is new to the recipe, add it to the
                        # mapping table
                        if RecipeIngredients.query.get((recipe_id, ing_id)) is None:
                            recipe_ingredient = RecipeIngredients(
                                recipe_id=recipe_id,
                                ingredient_id=ing_id,
                                unit_of_measure=ing["unit_of_measure"],
                                quantity=ing["quantity"],
                            )
                            db.session.add(recipe_ingredient)
                        # If not, just update the mapping table
                        else:
                            recipe_ingredient = RecipeIngredients.query.get(
                                (recipe_id, ing_id)
                            )
                            for k, v in ing.items():
                                setattr(recipe_ingredient, k, v)
            # If the ingredients do not contain enough information, return an error (400)
            except KeyError as e:
                jsonify(
                    {"error:": f"Attribute {e} does not exist in the payload."}
                ), 400
        # Update all the other attributes of the recipe
        for k, v in request.json.items():
            setattr(recipe, k, v)
        # Commit all the changes to the database
        db.session.merge(recipe)
        db.session.commit()
        return recipe_schema.jsonify(recipe), 200

    # If DELETE, delete the recipe
    elif request.method == "DELETE":
        # Get the recipe or return 404 Not Found
        recipe = db.get_or_404(Recipe, recipe_id)
        # Get all the ingredients associated with the recipe in the mapping table
        n_rows = RecipeIngredients.query.filter(
            RecipeIngredients.recipe_id == recipe_id
        ).delete()
        logger.debug(f"{n_rows} will be deleted from the `recipe_ingredients` table.")
        # Delete the recipe
        db.session.delete(recipe)
        # Apply the changes
        db.session.commit()
        # Return confirmation of deletion
        return f"Recipe {recipe_name} was successfully deleted."


@app.route("/ingredients", methods=["GET"])
def ingredients():
    """Get all the ingredients."""
    all_ingredients = Ingredient.query.all()
    result = ingredients_schema.dump(all_ingredients)
    return jsonify(result), 200


@app.route("/ingredients/<ingredient_name>/recipes", methods=["GET"])
def recipes_by_ingredient(ingredient_name: str):
    """Get all the recipes given an ingredient.

    Args:
        ingredient_name (str): Name of the ingredient
    """
    # Get the ingredient ID
    ing_id = hash_string(ingredient_name, 18)
    logger.debug(
        f"Cleaned Ingredient Name: {ingredient_name}, " f"Ingredient ID: {ing_id} ."
    )
    # Return 404 Not Found is the ing_id does not exist
    ingredient = db.get_or_404(Ingredient, ing_id)
    # Get all the recipe ID from the mapping table that are associate with the ingredient ID
    recipe_ids = db.select(RecipeIngredients.recipe_id).where(
        RecipeIngredients.ingredient_id == ing_id
    )
    # Return the recipes
    recipes = Recipe.query.filter(Recipe.id.in_(recipe_ids)).all()
    result = recipes_schema.dump(recipes)
    return jsonify(result), 200


if __name__ == "__main__":
    app.run(host="localhost", port=5000, debug=True)
